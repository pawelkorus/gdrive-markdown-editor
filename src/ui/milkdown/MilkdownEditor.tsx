import React, { memo } from 'react';
import { Editor, rootCtx, defaultValueCtx, editorViewOptionsCtx } from '@milkdown/core';
import { Milkdown, useEditor } from '@milkdown/react';
import { commonmark } from '@milkdown/preset-commonmark';
import { history } from '@milkdown/plugin-history';
import { listener, listenerCtx } from "@milkdown/plugin-listener";
import { cursor } from '@milkdown/plugin-cursor';
import { useSetProseState } from './dev-plugin/ProseStateProvider';
import { debounce } from 'lodash';
import { useGdriveEmbed } from './gdrive-plugin';
import { remarkPlugins } from './remark-plugin';

type MilkdownEditorProps = {
    content: string,
    readonly?: boolean,
    onContentUpdated?: (markdown:string) => void
}

export default memo(function(props:MilkdownEditorProps) {
    const gdriveEmbed = useGdriveEmbed();
    const setProseState = useSetProseState();
    
    useEditor((root) => 
        Editor.make()
                .config((ctx) => {
                    ctx.set(rootCtx, root);
                    ctx.set(defaultValueCtx, props.content)

                    ctx.update(editorViewOptionsCtx, (prev) => ({
                        ...prev,
                        editable: () => !props.readonly
                    }))

                    ctx.get(listenerCtx)
                    .markdownUpdated((_, doc) => {
                        props.onContentUpdated && props.onContentUpdated(doc);
                    })
                    .updated((_, doc) => {
                        const state = doc.toJSON();
                        setProseState(state);
                        debounce(setProseState, 100)(state);
                    });
                })
                .use(commonmark)
                .use(listener)
                .use(history)
                .use(cursor)
                .use(remarkPlugins)
                .use(gdriveEmbed.plugins)
        , [props.content]);
        
    return <Milkdown />;
});
