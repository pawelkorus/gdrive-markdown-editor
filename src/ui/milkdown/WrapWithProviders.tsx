import { MilkdownProvider } from '@milkdown/react'
import { ProsemirrorAdapterProvider } from '@prosemirror-adapter/react'
import React, { PropsWithChildren } from 'react'

type Props = PropsWithChildren<object>

const WrapWithProviders: React.FC<Props> = ({ children }) => {
  return (
    <MilkdownProvider>
      <ProsemirrorAdapterProvider>
        {children}
      </ProsemirrorAdapterProvider>
    </MilkdownProvider>
  )
}

export default WrapWithProviders
