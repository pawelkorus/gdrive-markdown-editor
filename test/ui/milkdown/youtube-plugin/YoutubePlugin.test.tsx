import { test, expect } from 'vitest'
import React from 'react'
import { render, act } from '@testing-library/react'
import MilkdownEditor from '../../../../src/ui/milkdown/MilkdownEditor'
import WrapWithProviders from '../../../../src/ui/milkdown/WrapWithProviders'

const SAMPLE_CONTENT = '::youtube{src="sampleyoutube"}'

test('should render youtube embed iframe', async () => {
  await act(async () => {
    render(<WrapWithProviders><MilkdownEditor content={SAMPLE_CONTENT} /></WrapWithProviders>)

    await new Promise(r => setTimeout(r, 4000))
  })

  expect(document.body).toMatchSnapshot()
})
