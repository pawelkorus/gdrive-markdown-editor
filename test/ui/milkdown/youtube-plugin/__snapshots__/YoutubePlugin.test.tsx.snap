// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render youtube embed iframe 1`] = `
<body>
  <div>
    <div
      data-milkdown-root="true"
    >
      <div
        class="milkdown"
      >
        <div
          class="ProseMirror editor"
          contenteditable="true"
          role="textbox"
          translate="no"
        >
          <div
            data-node-view-root="true"
          >
            <p
              class="text-center ratio ratio-16x9"
              data-type="youtube"
            >
              <iframe
                allowfullscreen=""
                src="https://www.youtube.com/embed/sampleyoutube"
              />
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
